docs/features/non-source-outputs.md

```diff
@@ -0,0 +1,141 @@
+
+```xml
+<ItemGroup>
+    <!-- Map all of the outputs of the razor source generator -->
+    <CompilerArtifactMapping Include="Microsoft.CodeAnalysis.Razor.SourceGenerators.SourceGenerator" Location="$(OutDir)\Razor" />
+
+    <!-- Map an individual artifact -->
+    <CompilerArtifactMapping Include="ArtifactName" Location="c:\dir" />
+<ItemGroup>
+```
+
+Would produce the following in the generated editor config:
+
+```ini
+compiler_artifact.Microsoft.CodeAnalysis.Razor.SourceGenerators.SourceGenerator.outputdir = bin\net8.0\Razor
+compiler_artifact.ArtifactName.outputdir = c:\dir
+```
+
+This means that generator authors are free to specify a mapping alongside their generator via the standard `NuGet` props and targets mechanism, and consumers of the generators are able to override the location in a consistent way.
+
+## MSBuild semantics
+
+Today, it's not possible to dynamically provide the list of outputs to an `MSBuild` target. This proposal allows the compiler to effectively emit an unbound number of files, none of which are tracked via `MSBuild` meaning that incremental compilation can be broken. (Note this is actually true today thanks to `EmitCompilerGeneratedFiles` but they don't have a bearing on the correctness of build).
+
+In order to correctly participate in the build process the compiler and targets are updated to emit a new file with the list of all files that it wrote as part of a compilation. A new target is added before `CoreCompile` that runs unconditionally and looks for the existence of the output list file; if the list is present and anything in the list is missing or out of date, the target touches a 'sentinel file'. The 'sentinel' file is passed as an input to `CoreCompile`. This ensures that anytime an output from the compiler is modified outside of the compilation process, `CoreCompile` will be correctly invoked and not skipped as being up to date. The 'sentinel' file effectively stands in for all of the outputs produced by the compiler in the `MSBuild` up to date checks.
+
+In order to decide if a file is modified the new target will require a task that can decide if a file has been modified since the previous build. This logic is fairly simple, as it just has to compare the timestamp of the output list file with the items listed within, but it's worth calling out that this does duplicate some of the `MSBuild` logic in the compiler tasks. The concept of dynamic inputs/outputs is proposed by `MSBuild` but not currently implemented. If this feature is added we could remove this duplication.
+
+## Open Questions
+
+- Naming: This proposal originally used the term `NonSource` but that's not particularly descriptive and perhaps odd in that it's not actually the mirror of a source output in that it's not going into the compilation. `Artifact` seems like a better term? Or simply `File`?
```

Comment:
You're right that `NonSource` doesn't really run off the tongue. I'm ambivalent between `File` and `Artifact`. The Roslyn API tends to use `File` here (`SourceFile`, `AdditionalFile`, etc ...) so I'd lean towards that but I don't feel strongly about this.

Reply:
AdditionalOutputFile is consistent and clear, if long.